lms/djangoapps/certificates/generation_handler.py

```diff
@@ -239,6 +245,122 @@ def _can_generate_certificate_common(user, course_key):
     return True
 
 
+def _set_allowlist_cert_status(user, course_key):
+    """
+    Determine the allowlist certificate status for this user, in this course run and update the cert.
+
+    This is used when a downloadable cert cannot be generated, but we want to provide more info about why it cannot
+    be generated.
+    """
+    if not _can_set_allowlist_cert_status(user, course_key):
+        return None
+
+    cert = GeneratedCertificate.certificate_for_student(user, course_key)
+    return _get_cert_status_common(user, course_key, cert)
+
+
+def _set_v2_cert_status(user, course_key):
+    """
+    Determine the V2 certificate status for this user, in this course run.
+
+    This is used when a downloadable cert cannot be generated, but we want to provide more info about why it cannot
+    be generated.
+    """
+    if not _can_set_v2_cert_status(user, course_key):
+        return None
+
+    cert = GeneratedCertificate.certificate_for_student(user, course_key)
+    status = _get_cert_status_common(user, course_key, cert)
+    if status is not None:
+        return status
+
+    course = _get_course(course_key)
+    course_grade = _get_course_grade(user, course)
+    if not course_grade.passed:
+        if cert is None:
+            cert = GeneratedCertificate.objects.create(user=user, course_id=course_key)
+        if cert.status != CertificateStatuses.notpassing:
+            cert.mark_notpassing(course_grade.percent)
+        return CertificateStatuses.notpassing
+
+    return None
+
+
+def _get_cert_status_common(user, course_key, cert):
+    """
+    Determine the certificate status for this user, in this course run.
+
+    This is used when a downloadable cert cannot be generated, but we want to provide more info about why it cannot
+    be generated.
+    """
+    if CertificateInvalidation.has_certificate_invalidation(user, course_key):
+        if cert is None:
+            cert = GeneratedCertificate.objects.create(user=user, course_id=course_key)
```

Comment:
Do we need to check if the cert is None here? The CertificateInvalidation entry has a FK to the GeneratedCertificate record. If `has_certificate_invalidation` is `True` then I think we can safely assume that the certificate exists for the learner?

Reply:
Good point. This is probably unnecessary as you said.

I'm slightly leaning toward leaving it in for consistency with the other checks, but I'll think about this before merging.